enhancements/oc/mirroring-manifests.md

```diff
@@ -0,0 +1,182 @@
+
+### Non-Goals
+
+This proposal does not specify the types of manifests that are created.
+Users who intend to audit the gathered manifests should extract the gathered types from the manifests themselves.
+
+## Proposal
+
+### Applying directly to the target cluster
+
+As of 4.2 and 4.3:
+
+```console
+$ oc adm release mirror \
+>   --from=quay.io/openshift-release-dev/ocp-release:4.3.0-x86_64 \
+>   --to=registry.example.com/your/repository \
+>   --to-release-image=registry.example.com/your/repository:your-tag \
+>   --apply-manifests
+```
+
+will mirror the 4.3.0 release image and other images referenced by the release image into the mirror repository at registry.example.com.
+
+This enhancement would also apply associated manifests to the cluster that `oc` connects to, with the manifest application beginning after the image mirror had completed.
+This enhancement would also allow `oc apply`'s `--force-conflicts` and `--overwrite` options to allow users to manage conflicts with resources that already existed in the target cluster.
+If `oc` encountered an error applying a manifest, it could optionally attempt to apply additional manifests before exiting non-zero, but would not be required to attempt any additional manifests.
+
+### Pushing to disk
+
+As of 4.3:
+
+```console
+$ oc adm release mirror --to-dir=mirror quay.io/openshift-release-dev/ocp-release:4.3.0-x86_64
+...
+Success
+Update image:  openshift/release:4.3.0
+
+To upload local images to a registry, run:
+
+    oc image mirror --from-dir=mirror file://openshift/release:4.3.0* REGISTRY/REPOSITORY
+```
+
+creates an output filesystem like:
+
+```console
+$ tree mirror
+mirror
+└── v2
+    └── openshift
+        └── release
+            ├── blobs
+            │   ├── sha256:005337dc1f6870934cf5efaf443786149e6a71041a91566ce1e16e78bce511a9
+            ...
+            │   └── sha256:ffc132fefac522ec38b1b22ecd67a5173565efb3e8929e47ef7e0b4ee7920adf
+            └── manifests
+                ├── 4.3.0 -> mirror/v2/openshift/release/manifests/sha256:3a516480dfd68e0f87f702b4d7bdd6f6a0acfdac5cd2e9767b838ceede34d70d
+                ...
+                ├── 4.3.0-thanos -> mirror/v2/openshift/release/manifests/sha256:42ab3f59d3e769e88af89c30304b9cc77a76b66960747e2e1a6f8db097420858
+                ├── sha256:00ec58112784d340179e045bc70399ac1bd509ae43c4411e934ced23d016b2a1
+                ...
+                └── sha256:ffa87c992c5a2c51290e6d67cb609f29893dd147c31cff639ff6785ae7a1cfe2
+
+5 directories, 514 files
+```
+
+that can be applied to the local registry with the given `oc image mirror` command.
+
+With this enhancement, the output format would gain an additional directory `manifests` as a sibling to the current `v2` containing manifest files that can be applied to the local cluster with:
+
+```console
+$ oc apply -Rf mirror/manifests
+```
```

Comment:
"Applying" the manifests to the local cluster stores them where CVO will check to see if it there are available locally and then not attempt to download - or some similar logic? IOW, what are the semantics of the "oc apply" and how are the local manifests then used?

Reply:
> "Applying" the manifests to the local cluster stores them where CVO will check to see if it there are available locally and then not attempt to download - or some similar logic? 

Yup.  It just slurps in the local YAML content and creates a ConfigMap in the target cluster.  The CVO running in that cluster can read ConfigMaps that live alongside it in the cluster, but obviously can't read them right off whatever disk the user is calling `oc adm release mirror ...` on.

Reply:
We need to apply all the manifests present in `mirror/kubernetes-manifests`. So irrespective of sub-directories present or not we need to apply all the manifests. I do not see a strong reason having the `-R`.  We should build the logic all going through all the sub directories and applying it by default.